.github/workflows/wpcloud.yml

```diff
@@ -47,17 +49,60 @@ jobs:
       - name: Build wpcomsh
         if: steps.changed.outputs.wpcomsh == 'true'
         run: |
+          find . -path ./.github -prune -o -type f -print | sort > /tmp/before.txt
           echo "::group::Installing and building wpcomsh"
           pnpm jetpack build -v --deps plugins/wpcomsh
           echo "::endgroup::"
 
+          # We only want to save the files that were actually created or changed.
+          # But we can't just list them for actions/cache/save, "Argument list too long".
+          # So instead we delete all the unchanged files so we can tell actions/cache/save
+          # to save everything that's left.
+          git -c core.quotepath=off diff --name-only | sort > /tmp/changed.txt
+          if [[ -s /tmp/changed.txt ]]; then
+            grep -F -x -v -f /tmp/changed.txt /tmp/before.txt > /tmp/remove.txt
+          else
+            cp /tmp/before.txt /tmp/remove.txt
+          fi
+          xargs -d '\n' rm < /tmp/remove.txt
+          find . -type d -empty -delete
+
+      - name: Save wpcomsh build cache
+        if: steps.changed.outputs.wpcomsh == 'true'
+        id: wpcomsh-build-cache-save
+        uses: actions/cache/save@v4
+        with:
+          path: |
+            .
+            !./.github/
+          key: ${{ github.sha }}
   deploy:
     name: Run PHPUnit on the WPCloud test site
     runs-on: ubuntu-latest
     needs: build
+    if: needs.build.changed.outputs.wpcomsh == 'true'
     steps:
+      - uses: actions/checkout@v4
+
+      - name: Restore wpcomsh build cache
+        id: wpcomsh-build-cache
+        if: needs.build.result == 'success'
```

Comment:
I think we can leave out this line, it's there in the E2E workflow because the build might be skipped when it's going to run against Atomic or something.
```suggestion
```